Panic in gapis

Environment: MBP
Version: main truck @ adf1817a26839e81b45c440785c67283f9033890 (May 26 2017)
Log:
panic: runtime error: index out of range
  panic: runtime error: index out of range

goroutine 3928 [running]:
github.com/google/gapid/gapis/resolve.(*CommandTreeResolvable).Resolve.func3(0xbbd41, 0x1)
        /Users/ggl/go/src/github.com/google/gapid/gapis/resolve/command_tree.go:290 +0xca
github.com/google/gapid/gapis/atom.(*Group).AddAtoms.func1(0xbbd42)
        /Users/ggl/go/src/github.com/google/gapid/gapis/atom/group.go:315 +0x72
github.com/google/gapid/gapis/atom.(*Group).AddAtoms(0xc44efae8c0, 0xc420113ef0, 0x7d0, 0xc463c95460, 0x0)
        /Users/ggl/go/src/github.com/google/gapid/gapis/atom/group.go:339 +0x164
github.com/google/gapid/gapis/atom.(*Group).AddAtoms(0xc44efae580, 0xc420113ef0, 0x7d0, 0xc463c95370, 0x0)
        /Users/ggl/go/src/github.com/google/gapid/gapis/atom/group.go:334 +0x404
github.com/google/gapid/gapis/atom.(*Group).AddAtoms(0xc44efae440, 0xc420113ef0, 0x7d0, 0xc463c95290, 0x0)
        /Users/ggl/go/src/github.com/google/gapid/gapis/atom/group.go:334 +0x404
github.com/google/gapid/gapis/atom.(*Group).AddAtoms(0xc44efae340, 0xc420113ef0, 0x7d0, 0xc44efae2c0, 0x0)
        /Users/ggl/go/src/github.com/google/gapid/gapis/atom/group.go:334 +0x404
github.com/google/gapid/gapis/atom.(*Group).AddAtoms(0xc47b2bc7c8, 0xc420113ef0, 0x7d0, 0xc47b2bc7c0, 0xbbd41)
        /Users/ggl/go/src/github.com/google/gapid/gapis/atom/group.go:334 +0x404
github.com/google/gapid/gapis/resolve.(*CommandTreeResolvable).Resolve(0xc44e879730, 0x6b36860, 0xc42a0e9410, 0x11c29348, 0xc44e879730, 0x1, 0x6b112e0)
        /Users/ggl/go/src/github.com/google/gapid/gapis/resolve/command_tree.go:293 +0x6e2
github.com/google/gapid/gapis/database.(*record).resolve(0xc42709e7b0, 0x6b367a0, 0xc4798675c0, 0xc45a032300, 0xc42a0e8de0)
        /Users/ggl/go/src/github.com/google/gapid/gapis/database/memory.go:72 +0x8a
github.com/google/gapid/gapis/database.(*memory).resolveLocked.func1(0xc42709e7b0, 0xc479867600, 0xc420334080)
        /Users/ggl/go/src/github.com/google/gapid/gapis/database/memory.go:144 +0x45
created by github.com/google/gapid/gapis/database.(*memory).resolveLocked
        /Users/ggl/go/src/github.com/google/gapid/gapis/database/memory.go:151 +0x4ff

--- a/gapis/atom/group.go
+++ b/gapis/atom/group.go
@@ -269,6 +269,15 @@ func (g Group) IterateBackwards(index uint64, cb func(childIdx uint64, item Grou
 // Attemping to call this function after atoms have been added may result in
 // panics!
 func (g *Group) AddGroup(start, end ID, name string) error {
+	if start > end {
+		return fmt.Errorf("sub-group start (%d) is greater than end (%v)", start, end)
+	}
+	if start < g.Range.Start {
+		return fmt.Errorf("sub-group start (%d) is earlier than group start (%v)", start, g.Range.Start)
+	}
+	if end > g.Range.End {
+		return fmt.Errorf("sub-group end (%d) is later than group end (%v)", end, g.Range.End)
+	}
 	r := Range{Start: start, End: end}
 	s, c := interval.Intersect(&g.Spans, r.Span())
 	if c == 0 {

--- a/gapis/resolve/command_tree.go
+++ b/gapis/resolve/command_tree.go
@@ -192,7 +192,7 @@ func (g *markerGrouper) process(ctx context.Context, id atom.ID, a atom.Atom, s 
 
 func (g *markerGrouper) flush(count uint64) {
 	for len(g.stack) > 0 {
-		g.pop(atom.ID(count))
+		g.pop(atom.ID(count) - 1)
 	}
 }
 
@@ -226,9 +226,9 @@ func (r *CommandTreeResolvable) Resolve(ctx context.Context) (interface{}, error
 	}
 
 	if p.GroupByContext {
-		var noContextId interface{} = nil
+		var noContextID interface{}
 		if p.IncludeNoContextGroups {
-			noContextId = gfxapi.ContextID{}
+			noContextID = gfxapi.ContextID{}
 		}
 		groupers = append(groupers, &runGrouper{f: func(a atom.Atom, s *gfxapi.State) (interface{}, string) {
 			if api := a.API(); api != nil {
@@ -236,7 +236,7 @@ func (r *CommandTreeResolvable) Resolve(ctx context.Context) (interface{}, error
 					return context.ID(), context.Name()
 				}
 			}
-			return noContextId, "No context"
+			return noContextID, "No context"
 		}})
 	}
 

